[add commit] 리뷰 사항 반영

parkcheayeon opened the pull request.

parkcheayeon committed:
키보드로 텍스트 입력받는 View, 검색버튼 PlantListFragment 위에 추가

parkcheayeon committed:
search view, 검색 버튼 리스너 등록, PlantListFragment, PlantAdapter 분석

Dukoo1 commented on app/src/main/res/layout/fragment_plant_list.xml:
```diff
@@ -23,10 +23,34 @@
+                android:textColor="?attr/colorPrimary"
+                android:text="@string/search_title"
+                app:backgroundTint="?attr/colorOnPrimary"
+                app:shapeAppearance="@style/ShapeAppearance.Sunflower.Button.Add"/>
```
기존 버튼 style을 쓰셔서 UI 통일성 측면에서 좋겠네요.
근데 이름으로 추측해보면 이 버튼은 Add 전용을 쓰려는 의도같아요.
추후에 Add버튼을 변경하면, Search버튼까지 바뀌는 부작용이 있을 수 있겠네요.
만약 협업중이라고 가정하면,

1. `ShapeAppearance.Sunflower.Button.Add` 이름을 `ShapeAppearance.Sunflower.Button`으로 바꾸자고 합의후에 공용 style로 쓴다던지,
2. 아님 Search용 style을 따로 만드는 방법이 있겠죠. (내용은 같을 수도 있고, 다를 수도 있고요)

우리는 후자로 함 해볼까요? Search버튼은 네귀퉁이 모두 동그랗게.

Dukoo1 commented on app/src/main/res/layout/fragment_plant_list.xml:
```diff
@@ -23,10 +23,34 @@
             android:id="@+id/plant_list"
             android:layout_width="match_parent"
             android:layout_height="match_parent"
+        android:layout_marginTop="56dp"
```
Search view 아래 list view를 놓으려는 의도같은데요.
Search view height가 바뀌면 여기 숫자도 항상 같이 바꿔줘야하죠.
marginTop 빼고, layout구조를 좀 바꿔서, Search view 아래 list view가 놓이게 해보시겠어요?

parkcheayeon replied:
커밋내용에는 빠졌지만 FrameLayout을 ConstraintLayout로 변경하였습니다

Dukoo1 commented on app/src/main/java/com/google/samples/apps/sunflower/PlantListFragment.kt:
```diff
@@ -122,19 +126,45 @@ class PlantListFragment : Fragment() {
-            } else {
-                setGrowZoneNumber(9)
+    private fun initSearchView() {
+        binding.searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
```
전 일반적 EditView를 쓰실줄 알았는데, 더 고급진 SearchView를 쓰셨네요. 👍

참고로, 리스너에서 리턴값을 통해 기본액션을 할지, 별도 액션을 알아서 정하는건
안드로이드 여기저기에서 두루 쓰이는 방법입니다.
View마다 다를 수 있으니 주석을 잘 읽어봐야겠죠.

Dukoo1 commented on app/src/main/java/com/google/samples/apps/sunflower/adapters/PlantAdapter.kt:
```diff
@@ -41,17 +41,21 @@ class PlantAdapter : ListAdapter<Plant, RecyclerView.ViewHolder>(PlantDiffCallba
         (holder as PlantViewHolder).bind(plant)
+        // 여기서는 타입캐스팅을 하는 이유는?
+        // RecyclerView.ViewHolder는 모든 ViewHolder의 상위 타입이기 때문
+        // Adapter의 타입을 PlantViewHolder를 명시하면 되는거 아닐까?
```
좋은 생각이에요, 시도해보세요.
generic을 잘쓰면 run-time type check를 compile-time type check로 바꿀 수 있으니 실수도 줄어들죠.

Dukoo1 commented on app/src/main/res/layout/fragment_plant_list.xml:
```diff
@@ -23,10 +23,34 @@
+                android:id="@+id/search_view"
+                android:layout_width="0dp"
+                android:layout_height="wrap_content"
+                android:layout_weight="1"
```
아무것도 입력안했을땐 "Enter plant"라고 보여주도록 해보시겠어요?
(hint가 힌트입니다)

parkcheayeon committed:
PlantListViewModel 분석, 검색기능 추가

Dukoo1 commented on app/src/main/java/com/google/samples/apps/sunflower/viewmodels/PlantListViewModel.kt:
```diff
@@ -38,53 +38,100 @@ import javax.inject.Inject
+        MutableStateFlow(savedStateHandle.get(GROW_ZONE_SAVED_STATE_KEY) ?: NO_GROW_ZONE)
+    // MutableStateFlow => 상태 변경 시 이를 실시간으로 제공하고, 읽기와 쓰기가 모두 가능하다
+    // 초기값이 항상 필요 왜? 항상 현재 상태를 제공 해야 하기 때문에
+    // 그럼 그냥 flow랑은 뭐가 다를까? flow는 DB에서 데이터를 한 번 가져오는 작업에 주로 사용한다
```
Flow와 StateFlow는 좀 어렵긴한데 많은 차이가 있습니다.
대략 Flow -> StateFlow -> MutableStateFlow 이렇게 상속관계가 있는데요.
Flow는 자체상태유지가 없고, collector/observer 가 구독을 해줘야 flow가 실행됩니다.(cold라고 부릅니다.)
StateFlow는 이름에서 알수있듯 상태가 있고(async완료 전에도 리턴할게 필요하니까 초기값도 필요하죠), collector/observer가 없을수도 하나일수도 여러개가 붙어서 read할수도 있습니다(hot이라 부릅니다). Mutable까지 붙으면 여러군데서 write도 할수있구요.

Dukoo1 commented on app/src/main/java/com/google/samples/apps/sunflower/viewmodels/PlantListViewModel.kt:
```diff
@@ -38,53 +38,100 @@ import javax.inject.Inject
+    // 그럼 그냥 flow랑은 뭐가 다를까? flow는 DB에서 데이터를 한 번 가져오는 작업에 주로 사용한다
+    // growZone 값 -1 또는 9
+
+    // flatMapLatest => growZone 값이 변경될 때마다 실행
```
각종 Collections는 map, mapLatest, flarMapLatest 등의 map 함수를 가지고 있는데요.

여기서 map은 HashMap의 map보다는 변환(mapping)의 의미라고 보시면 되겠습니다. 겉의 Flow형태는 유지하되 알맹이가 Int 타입에서 Plant타입으로 변환되었죠.

Dukoo1 commented on app/src/main/java/com/google/samples/apps/sunflower/viewmodels/PlantListViewModel.kt:
```diff
@@ -38,53 +38,100 @@ import javax.inject.Inject
 @HiltViewModel
 class PlantListViewModel @Inject internal constructor
     (plantRepository: PlantRepository, private val savedStateHandle: SavedStateHandle) : ViewModel() {
+        // SavedStateHandle => ViewModel의 데이터를 구성 변경이나 프로세스 종료 후에도 안전하게 저장하고 복원할 수 있다 (키-값 형태로 저장)
```
ViewModel에 단순 멤버변수로만 넣어둬도, configuration change에 살아남습니다.
하지만 여기선 LowMemoryKiller로 죽은 프로세스가 복구될때까지도 유지해주게 했네요.
프로세스가 죽었던걸 사용자에게 들키지않게하려는 고급스킬입니다.ㅎ
https://developer.android.com/topic/libraries/architecture/viewmodel/viewmodel-savedstate

Dukoo1 commented on app/src/main/java/com/google/samples/apps/sunflower/viewmodels/PlantListViewModel.kt:
```diff
@@ -38,53 +38,100 @@ import javax.inject.Inject
+    private val keyWord: MutableStateFlow<String> =
+        MutableStateFlow(savedStateHandle.get(KEYWORD_SAVED_STATE_KEY) ?: NO_KEYWORD)
 
+    private val growZone: MutableStateFlow<Int> =
```
MutableStateFlow는 class가 아니라 interface인데(https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-mutable-state-flow/)
어떻게 이렇게 객체를 생성할수있을까요?

사실 뒤에건 함수입니다.
https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-mutable-state-flow.html

코틀린에서 자주 쓰는 패턴인데요.
인터페이스 이름과 함수 이름을 같게 지어요.
헷갈린다고 볼수도 있고, 더 읽기쉽다고 볼수도 있구요.(이제보니 문서url도 헷갈리게 만들어놨네요 ㅎㅎ)

parkcheayeon committed:
PR 리뷰 사항 반영
1. search button 전용 스타일 추가 후 변경
2. SearchView 리스너 리턴 값 변경
3. SearchView 속성 추가
4. PlantAdapter 뷰홀더 타입 변경

Dukoo1 commented on app/src/main/java/com/google/samples/apps/sunflower/PlantListFragment.kt:
```diff
@@ -151,18 +151,21 @@ class PlantListFragment : Fragment() {
             override fun onQueryTextSubmit(query: String?): Boolean {
-                if (query != null) {
+                query?.let {
                     Log.e("search", "text submit : $query")
```
let 람다 안에선 $query는 뭘로 바꿀 수 있을까요?

Dukoo1 commented on app/src/main/res/layout/fragment_plant_list.xml:
```diff
@@ -19,46 +18,53 @@
     xmlns:tools="http://schemas.android.com/tools">
 
-    <FrameLayout
+    <androidx.constraintlayout.widget.ConstraintLayout
```
오~ vertical LinearLayout으로 감싸실줄 알았는데, ConstraintLayout을 쓰셨군요.

LinearLayout은 사용이 간단한게 장점이지만 depth가 늘어나고.
ConstraintLayout을 여러 자식들을 depth를 줄여줍니다. flat하다고도 부르죠.
깊은 depth를 가진 view tree일수록 그릴때 더 느리거든요.(물론 현재 이 정도 depth는 괜찮습니다)

constraint를 잘써서 SearchView와 MaterialButton을 ConstraintLayout 바로 아래 depth로 바꿔보세요~

Dukoo1 commented on app/src/main/java/com/google/samples/apps/sunflower/PlantListFragment.kt:
```diff
@@ -152,7 +152,7 @@ class PlantListFragment : Fragment() {
         binding.searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
             override fun onQueryTextSubmit(query: String?): Boolean {
                 query?.let {
-                    Log.e("search", "text submit : $query")
+                    Log.e("search", "text submit : $it")
                 }
```
정답입니다.
간단할땐 it으로 하기도 하고,
람다 속이 복잡해서 it보다 더 가독성 좋은 이름으로 하고 싶다면, 이름을 정해줄 수도 있습니다.
```
                query?.let { qqq ->
                    ...복잡한 코드
                    Log.e("search", "text submit : $qqq")
                }
```
